### PR TITLE
Blaze Campaign: web view refactoring

### PR DESCRIPTION
**Pull request by zwarm**
This PR implements refactoring of the campaign detail web view to use UI states. The first few commits leaned more towards a boolean approach, while the remainder of the commits follow the uiState approach.

Changes include
-The view moves from `Preparing` to `Prepared` to `Loaded` states. The loading indicator is shown while in the preparing and prepared states. It gets removed on call from the webview that the page has loaded.

-Show a loading indicator, the view will now show an error view if account details are not present, there is no network connectivity, or the webview throws an error. I included a "retry" button, not sure if it should be there for non network connectivity errors. wdyt? the error view itself is based off of the No Campaigns view. FYI @osullivanchris 

-`CampaignDetailViewModel` implements ScopedViewModel and posts the navigationEvents objects in a background thread. However the uiStates ares directly observed in the UI via Compose. This means that the UI updates will still happen on the main thread, as Compose handles the observation and recomposition on the main thread by default.

Loading Before | Loading After
-- | --
<video src="https://github.com/wordpress-mobile/WordPress-Android/assets/506707/14af0e5e-d132-4678-a40a-6fd4eb266a3a"> | <video src="https://github.com/wordpress-mobile/WordPress-Android/assets/506707/c2dae46d-e3d6-4aa9-b867-dc5f5c8a1285">

No Network Dark | No Network Light
-- | --
 | 

Error Dark | Error Light
-- | --
 | 

### To test:
**Setup**
- Install and launch the app
- Navigate to Me -> App Settings -> Privacy Setting
- Enable collection information
- Navigate to Me -> App Settings -> Debug Setting
- Enable Blaze
- Restart the app

**Viewing Campaign Details**
- Navigate to a site that has blaze campaigns (use the test creds or create a blaze campaign first)
- Navigate to My Site tab
- Tap on the Campaign Detail section within the Blaze Card to launch the Campaign Detail View
- ✅ Verify the Loading indicator is shown while the page is loading
- ✅ Verify the Loading indicator goes away when the page is loaded

**Error view shows when there is no internet**
- Navigate to a site that has blaze campaigns (use the test creds or create a blaze campaign first)
- Navigate to My Site tab > Menu tab
- Tap on the Blaze menu item
- Verify the campaign list view has launched
- **STOP** - Turn on airplane mode
- Tap on any campaign within the campaign list
- ✅ Verify the Campaign Detail view is launched and the No Internet Error view is shown

**Error view shows when there is no account creds/or the webview is throws an error**
- Download the branch
- In fun `validateAndPostErrorIfNeeded` comment out the if statement and the return false
- Build, install, and run
- Navigate to My SIte tab 
- Tap on the Campaign Detail section within the Blaze Card to launch the Campaign Detail View
✅ Verify the generic error view is shown  

## Regression Notes
1. Potential unintended areas of impact
The blaze campaign detail view does not work as expected

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing + `CampaignDetailViewModelTest`

3. What automated tests I added (or what prevented me from doing so)
Updated/added tests to `CampaignDetailViewModelTest`

PR submission checklist:
- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [X] Portrait and landscape orientations.
- [X] Light and dark modes.
- [X] Fonts: Larger, smaller and bold text. 
- [X] High contrast.
- [X] Talkback.
- [X] Languages with large words or with letters/accents not frequently used in English.
- [X] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [X] Large and small screen sizes. (Tablet and smaller phones)
- [X] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/blaze/blazecampaigns/campaigndetail/CampaignDetailViewModel.kt [hunk 1]
```diff
@@ -104,12 +109,12 @@ class CampaignDetailViewModel @Inject constructor(
     }
 
     private fun postScreenState(state: CampaignDetailUIModel) {
-        viewModelScope.launch {
+        viewModelScope.launch(bgDispatcher) {
```
Comment (AjeshRPai) [inline WordPress/src/main/java/org/wordpress/android/ui/blaze/blazecampaigns/campaigndetail/CampaignDetailViewModel.kt]: Hey @zwarm 
I think just calling `launch` is enough here as we have already extended `ScopedViewModel`

File: WordPress/src/main/java/org/wordpress/android/ui/blaze/blazecampaigns/campaigndetail/CampaignDetailMapper.kt [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    fun toGenericError(buttonClick: () -> Unit) = CampaignDetailUiState.Error(
+        title = UiString.UiStringRes(R.string.campaign_detail_error_title),
+        description = UiString.UiStringRes(R.string.campaign_detail_error_description),
+        button = CampaignDetailUiState.Error.ErrorButton(
+            text = UiString.UiStringRes(R.string.campaign_detail_error_button_text),
+            click = buttonClick
+        )
+    )
+}
```
Comment (AjeshRPai) [inline WordPress/src/main/java/org/wordpress/android/ui/blaze/blazecampaigns/campaigndetail/CampaignDetailMapper.kt:23]: Hey @zwarm 
I like using a mapper for the prepared status 👍🏼 .
❓But do you think we need the mapper for the error state?
I tried this on my PR as well https://github.com/wordpress-mobile/WordPress-Android/pull/18842/commits. Once I did this, I felt that in the scenario of an error, we couldn't effectively test the type of error(Generic/No network) on the UI state. Then I moved the error types as a subclass of the error in this [commit ](https://github.com/wordpress-mobile/WordPress-Android/pull/18842/commits/056bb36c280085f7eed4462d9cff8213f6bdc209)

The benefit of dropping the `toNoNetworkError` and `toGenericError` and using data class to represent `NoNetworkError` and `GenericError` is that we can assert which type of error is in the UI state. See this [line](https://github.com/wordpress-mobile/WordPress-Android/pull/18842/files#diff-36428c2690917ef37687ab8574ab294a88ad37acd33f80f379c6cbe8d6723682R126). I don't have a strong opinion on both of this approach.  

One other approach we can do is to use the mapper function itself and use an error type value enum. Then assert on the ui test that the error type is  what we want. 

Would love to hear your thoughts

**Review by AjeshRPai — APPROVED** [review]
Hey @zwarm 

I have tested this PR according to the instructions. Everything looks good to me 👍🏼 . I can see the loading state/error state and the blaze flow is working as before 👍🏼 

Approving but not merging so that you can have a look at my ❓ .

File: WordPress/src/main/java/org/wordpress/android/ui/blaze/blazecampaigns/campaigndetail/CampaignDetailViewModel.kt (hunk 1 above)
Comment (zwarm) [reply WordPress/src/main/java/org/wordpress/android/ui/blaze/blazecampaigns/campaigndetail/CampaignDetailViewModel.kt]: You're right. In commit [e4d7918](https://github.com/wordpress-mobile/WordPress-Android/pull/18844/commits/e4d7918ffac936b066584e4174fbacb7c18b1b4f)

File: WordPress/src/main/java/org/wordpress/android/ui/blaze/blazecampaigns/campaigndetail/CampaignDetailMapper.kt (hunk 2 above)
Comment (zwarm) [reply WordPress/src/main/java/org/wordpress/android/ui/blaze/blazecampaigns/campaigndetail/CampaignDetailMapper.kt:23]: > Then I moved the error types as a subclass of the error

I'm on board with this idea, definitely much easier to validate if we have the correct error view showing. I also thought of the same `enum` solution, but felt the data class was the way to go in this instance. I kept the mapper functions `toNoNetworkError` and `toGenericError`. They are simple to mock for testing.

Nice refactor!

**Comment by osullivanchris** [comment]
I saw #18855  first so I left my feedback on that one. Looks like same applies to both.

**Pull request merged**